src/common/guac_surface.c

```diff
@@ -77,6 +79,446 @@
+
+                }
+
+                /* Exclude the remaining part of the dirty rectangle
+                 * which is completely covered by the lossy dirty rectangle. */
+                current->flushed = 1;
+
+            }
+            current++;
+
+        }
+
+        /* Flush the lossy bitmap */
+        __guac_common_surface_flush_to_bitmap_impl(surface,
+                                                 &surface->lossy_dirty_rect, 1);
+
+        /* Flag this area as lossy so it can be moved back to the
+         * dirty rect and refreshed normally when refreshed less frequently. */
+        int x = surface->lossy_dirty_rect.x;
+        int y = surface->lossy_dirty_rect.y;
+        int w = (x + surface->lossy_dirty_rect.width) / GUAC_COMMON_SURFACE_STAT_RECT;
+        int h = (y + surface->lossy_dirty_rect.height) / GUAC_COMMON_SURFACE_STAT_RECT;
+        x /= GUAC_COMMON_SURFACE_STAT_RECT;
+        y /= GUAC_COMMON_SURFACE_STAT_RECT;
+
+        for (int j = y; j <= h; j++) {
+            for (int i = x; i <= w; i++) {
+                surface->lossy_rect[j][i] = 1;
+            }
+        }
+
+        /* Clear the lossy dirty flag. */
+        surface->lossy_dirty = 0;
+    }
+
+}
+
+/**
+ * Calculate the current average update frequency for a given area on the
+ * surface.
+ *
+ * @param surface
+ *     The surface on which the update frequency will be calculated.
+ *
+ * @param x
+ *     The x coordinate for the area.
+ *
+ * @param y
+ *     The y coordinate for the area.
+ *
+ * @param w
+ *     The area width.
+ *
+ * @param h
+ *     The area height.
+ *
+ * @return
+ *     The average update frequency.
+ */
+static double __guac_common_surface_calculate_update_freq(
+                                                   guac_common_surface* surface,
+                                                   int x, int y, int w, int h)
+{
+    int i, j;
+    double total = 0;
+    int num = 0;
+
+    x =  (x / GUAC_COMMON_SURFACE_STAT_RECT) * GUAC_COMMON_SURFACE_STAT_RECT;
+    y =  (y / GUAC_COMMON_SURFACE_STAT_RECT) * GUAC_COMMON_SURFACE_STAT_RECT;
+
+    /* Iterate over all the statistics rectangles for the area
+     * and calculate the average update frequency. */
+    for (j = y; j <= y + h; j += GUAC_COMMON_SURFACE_STAT_RECT) {
+        for (i = x; i <= x + w; i += GUAC_COMMON_SURFACE_STAT_RECT) {
+
+            total += __guac_common_surface_stat_rect(surface, i, j)->freq;
+            num++;
+
+        }
+    }
+
+    /* Calculate the average. */
+    if (num) {
+        return total / num;
+    }
+    else {
+        return 0;
+    }
+}
+
+/**
+ * Update the statistics rectangles for the surface and re-calculate the update
+ * frequencies.
+ *
+ * Any areas of the surface which have not been updated within a given threshold
+ * will be moved from the lossy to the normal refresh path.
+ *
+ * Code inspiration comes from the QEMU VNC display driver:
+ * https://github.com/qemu/qemu/blob/master/ui/vnc.c
```

Comment:
Nope - can't legally do this. If the code here were only inspired by it, then yes, but much of this code is copied verbatim from the referenced file, which is a violation of their license. Including this code without following the terms of their license infringes on their copyright:

```
/*
 * QEMU VNC display driver
 *
 * Copyright (C) 2006 Anthony Liguori <anthony@codemonkey.ws>
 * Copyright (C) 2006 Fabrice Bellard
 * Copyright (C) 2009 Red Hat, Inc
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
```

Providing a link to the code is nice, but not sufficient to satisfy the license terms. The key provision here is:

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

The portions here are substantial, and we need to include the copyright/permission notice if we're going to use them.